x-pack/plugins/apm/public/hooks/useFetcher.tsx

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { flatten, isObject } from 'lodash';
+import { useEffect, useState } from 'react';
+
+export enum FETCH_STATUS {
+  LOADING = 'loading',
+  SUCCESS = 'success',
+  FAILURE = 'failure'
+}
+
+export function useFetcher<Args extends any[], Response>(
+  fn: (...args: Args) => Promise<Response>,
+  fnArgs: Args
+) {
+  const [status, setStatus] = useState<FETCH_STATUS | null>(null);
+  const [data, setData] = useState<Response | null>(null);
+  const [error, setError] = useState<Error | null>(null);
+  let didCancel: boolean;
+
+  async function fetchData() {
+    setStatus(FETCH_STATUS.LOADING);
+    setError(null);
+    try {
+      const fnData = await fn(...fnArgs);
+      if (!didCancel) {
+        setData(fnData);
+        setStatus(FETCH_STATUS.SUCCESS);
+      }
+    } catch (e) {
+      setError(e);
+      setStatus(FETCH_STATUS.FAILURE);
+    }
+  }
+
+  // support objects as arguments. Convert object to flat array and pass the key/values as a flat array
```

Comment:
OK I understand this better now. I still think we can find something slightly better but I don't have great ideas yet.

Only thing I can think of right now is to only accept a single `options` object that is passed onto the fetching function (rather than an array of arguments). We would explicitly just flatten the keys/values from that one options object, so that it's clearer that we are doing it only for that reason and not just generally flattening and using any arguments we receive ... do we ever currently pass in more than that one object as an argument to one of these functions anywhere?

Reply:
> Only thing I can think of right now is to only accept a single options object that is passed onto the fetching function (rather than an array of arguments).

I think this is a good idea. I did the other approach to keep it more versatile but I like the clearer intent.

> do we ever currently pass in more than that one object as an argument to one of these functions anywhere

Not as far as I can see. Sometimes we don't pass anything at all but that should work as well.